File: sdk/python/agent_eval/transport.py
"""传输层 - 异步 batch 上报到平台"""
import atexit
import json
import logging
import threading
import time
from typing import Any, Dict, List, Optional
from urllib.request import Request, urlopen
from urllib.error import URLError

logger = logging.getLogger("agent_eval.transport")


class BatchTransport:
    """
    异步 batch HTTP 传输。
    - 后台线程定时 flush（默认 5 秒）
    - 缓冲区达到阈值时立即 flush（默认 100 条）
    - 失败时指数退避重试（最多 3 次）
    - 所有异常静默捕获，不影响业务
    """

    def __init__(
        self,
        base_url: str,
        api_key: str,
        flush_interval: float = 5.0,
        max_batch_size: int = 100,
        max_retries: int = 3,
    ):
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.flush_interval = flush_interval
        self.max_batch_size = max_batch_size
        self.max_retries = max_retries

        self._buffer: List[Dict[str, Any]] = []
        self._lock = threading.Lock()
        self._running = True

        # 后台 flush 线程
        self._thread = threading.Thread(target=self._flush_loop, daemon=True)
        self._thread.start()

        # 退出时最后 flush
        atexit.register(self.shutdown)

    def enqueue(self, event: Dict[str, Any]) -> None:
        """将事件加入缓冲区"""
        try:
            batch = None
            with self._lock:
                self._buffer.append(event)
                if len(self._buffer) >= self.max_batch_size:
                    batch = self._buffer[:]
                    self._buffer.clear()
            if batch:
                self._send_batch(batch)
        except Exception:
            pass

    def _flush_loop(self) -> None:
        """后台线程：定时 flush"""
        while self._running:
            time.sleep(self.flush_interval)
            self._flush()

    def _flush(self) -> None:
        """执行一次 flush"""
        try:
            with self._lock:
                if not self._buffer:
                    return
                batch = self._buffer[:]
                self._buffer.clear()
            self._send_batch(batch)
        except Exception:
            pass

    def _send_batch(self, batch: List[Dict[str, Any]]) -> None:
        """发送一批事件到 POST /api/ingest"""
        if not batch:
            return

        url = f"{self.base_url}/api/ingest"
        payload = json.dumps({"events": batch}).encode("utf-8")
        headers = {
            "Content-Type": "application/json",
            "X-API-Key": self.api_key,
        }

        for attempt in range(self.max_retries):
            try:
                req = Request(url, data=payload, headers=headers, method="POST")
                with urlopen(req, timeout=10) as resp:
                    if resp.status < 300:
                        logger.debug("Batch sent: %d events", len(batch))
                        return
            except URLError as e:
                logger.debug("Send attempt %d failed: %s", attempt + 1, e)
                if attempt < self.max_retries - 1:
                    time.sleep(2 ** attempt)
            except Exception:
                break

        logger.warning("Failed to send batch of %d events after %d retries", len(batch), self.max_retries)

    def send_score(self, score_data: Dict[str, Any]) -> None:
        """直接发送 Score 到 POST /api/scores"""
        try:
            url = f"{self.base_url}/api/scores"
            payload = json.dumps(score_data).encode("utf-8")
            headers = {
                "Content-Type": "application/json",
                "X-API-Key": self.api_key,
            }
            req = Request(url, data=payload, headers=headers, method="POST")
            with urlopen(req, timeout=10):
                pass
        except Exception as e:
            logger.debug("Score send failed: %s", e)

    def shutdown(self) -> None:
        """关闭传输层，最后 flush"""
        self._running = False
        self._flush()

File: sdk/typescript/src/transport.ts
import type { TransportEvent, ScoreOptions } from './types';

/**
 * 批量传输层 — 负责将事件批量发送到平台 API。
 * 支持定时 flush、阈值 flush、指数退避重试。
 */
export class BatchTransport {
  private baseUrl: string;
  private apiKey: string;
  private flushInterval: number;
  private maxBatchSize: number;
  private buffer: TransportEvent[] = [];
  private timer: ReturnType<typeof setInterval> | null = null;
  private isShutdown = false;
  private maxRetries = 3;

  constructor(
    baseUrl: string,
    apiKey: string,
    flushInterval: number = 5000,
    maxBatchSize: number = 100,
  ) {
    this.baseUrl = baseUrl.replace(/\/+$/, '');
    this.apiKey = apiKey;
    this.flushInterval = flushInterval;
    this.maxBatchSize = maxBatchSize;

    // 启动定时 flush
    this.timer = setInterval(() => this.flush(), this.flushInterval);

    // Node.js 进程退出前 flush
    if (typeof process !== 'undefined' && process.on) {
      process.on('beforeExit', () => this.shutdown());
    }
  }

  /** 将事件放入缓冲区，达到阈值时自动 flush */
  enqueue(event: TransportEvent): void {
    if (this.isShutdown) return;
    this.buffer.push(event);
    if (this.buffer.length >= this.maxBatchSize) {
      this.flush();
    }
  }

  /** 立即上报 Score（不走批量缓冲） */
  async sendScore(score: ScoreOptions): Promise<void> {
    const payload = {
      trace_id: score.traceId,
      name: score.name,
      value: score.value,
      string_value: score.stringValue,
      comment: score.comment,
      source: score.source || 'sdk',
      span_id: score.spanId,
    };
    try {
      await this.post('/api/scores', payload);
    } catch (err) {
      // 静默处理，不影响业务
      if (typeof console !== 'undefined') {
        console.warn('[agent-eval-sdk] Score send failed:', err);
      }
    }
  }

  /** 手动 flush 缓冲区 */
  async flush(): Promise<void> {
    if (this.buffer.length === 0) return;
    const batch = this.buffer.splice(0);
    await this.sendBatch(batch);
  }

  /** 关闭传输层 */
  async shutdown(): Promise<void> {
    if (this.isShutdown) return;
    this.isShutdown = true;
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
    await this.flush();
  }

  private async sendBatch(events: TransportEvent[]): Promise<void> {
    if (events.length === 0) return;
    for (let attempt = 0; attempt < this.maxRetries; attempt++) {
      try {
        await this.post('/api/ingest', { events });
        return;
      } catch (err) {
        if (attempt < this.maxRetries - 1) {
          const delay = Math.pow(2, attempt) * 1000;
          await new Promise((resolve) => setTimeout(resolve, delay));
        } else if (typeof console !== 'undefined') {
          console.warn('[agent-eval-sdk] Batch send failed after retries:', err);
        }
      }
    }
  }

  private async post(path: string, body: unknown): Promise<unknown> {
    const url = `${this.baseUrl}${path}`;
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };
    if (this.apiKey) {
      headers['X-API-Key'] = this.apiKey;
    }

    // 支持 Node.js fetch (18+) 和 globalThis.fetch
    const res = await fetch(url, {
      method: 'POST',
      headers,
      body: JSON.stringify(body),
    });
    if (!res.ok) {
      throw new Error(`HTTP ${res.status}: ${res.statusText}`);
    }
    return res.json();
  }
}
